src/main/java/com/syntaxphoenix/spigot/smoothtimber/listener/BlockBreakListener.java

```diff
@@ -82,76 +76,18 @@ public void onBlockBreak(final BlockBreakEvent event) {
-                                final int amount = maxItems <= 1 ? maxItems : generateAmount(maxItems);
-                                if (animated) {
-                                    final Entity entity = change.toFallingBlock(block);
-                                    entity.setMetadata("STAnimate", new FixedMetadataValue(plugin, amount));
-                                    if (collect) {
-                                        entity.setMetadata("STCollect",
-                                            new FixedMetadataValue(plugin, player.getUniqueId().toString()));
-                                    }
-                                    continue;
-                                }
-                                if (collect) {
-                                    final ItemStack stack = change.getItemFromBlock(block);
-                                    stack.setAmount(amount);
-                                    player.getInventory().addItem(stack);
-                                    continue;
-                                }
-                                change.dropItemByBlock(block, amount);
-                            }
-                        }
-                        Bukkit.getScheduler().runTaskAsynchronously(PluginUtils.MAIN, () -> SmoothTimber.triggerFallEvent(event));
-                    }
+            final int maxItems = CutterConfig.ENABLE_LUCK ? change.getMaxDropCount(tool) : 1;
+            final ArrayList<Location> woodBlocks = new ArrayList<>();
 
-                    private int generateAmount(final int max) {
-                        int drop = 1;
-                        final float more = 1f / (max + 1);
-                        float previous = more * 2f;
-                        float next = more * 3f;
-                        final float chance = generator.nextFloat() * (float) CutterConfig.LUCK_MULTIPLIER;
-                        while (true) {
-                            if (previous < chance && chance > next) {
-                                drop++;
-                                previous = next;
-                                next += more;
-                            } else {
-                                if (previous < chance && chance < next) {
-                                    drop++;
-                                }
-                                break;
-                            }
-                        }
-                        return Math.min(drop, 64);
-                    }
-
-                });
-            });
+            BlockBreakTask blockBreakTask = new BlockBreakTask(player, location, change, tool, woodBlocks, maxItems, generator);
+            BlockBreakEventTask blockBreakEventTask = new BlockBreakEventTask(player, location, change, tool, woodBlocks, maxItems);
+            if (SmoothTimber.IS_FOLIA) {
+                Bukkit.getServer().getAsyncScheduler().runNow(PluginUtils.MAIN, value -> blockBreakEventTask.run());
```

Comment:
I just noticed that this won't even work.
Normally if the Chop Event of SmoothTimber is cancelled then the whole block break process will be aborted.
However your "blockBreakEventTask" is executed in parallel with the actual "blockBreakTask" which makes no sense.
The event has to be triggered and be checked before the "blockBreakTask" is scheduled and if the event that was triggered by the task was cancelled it should not schedule the "blockBreakTask".

As I said try to prevent unnecessary changes like these, it just makes more work to check through your PR and ensure that everything is working as intended in the exact same way as before.